File: DesignPattern/abtract_factory_design_pattern_table_chair.cpp
#include <iostream>
using namespace std;
enum MaterialType{
    WOOD, PLASTIC
};
//Abstract Chair
class Chair{
    public:
    virtual void CreateChair() = 0;
};
//Abstract Table
class Table{
    public:
    virtual void CreateTable() = 0;
};


class WoodenChair:public Chair
{
    void CreateChair(){
        cout<< "Wooden Chair"<<endl;
    }
};
class PlasticChair:public Chair
{
    void CreateChair(){
        cout<< "Plastic Chair"<<endl;
    }
};


class WoodenTable:public Table{
    void CreateTable()    {
        cout<<"Wooden table"<<endl;
    }
};
class PlasticTable:public Table{
    void CreateTable()    {
        cout<<"Plastic table"<<endl;
    }
};

//Abstract Factory
class AbstractFurnitureFactory{
public:
    virtual Table* CreateTable() = 0;
    virtual Chair* CreateChair() = 0;
    static AbstractFurnitureFactory* GetFactory(MaterialType type);
};

class WoodenFuniture:public AbstractFurnitureFactory{
    Table* CreateTable(){
        return new WoodenTable();
    }
    Chair* CreateChair(){
        return new WoodenChair();
    }
};
class PlasticFuniture:public AbstractFurnitureFactory{
    Table* CreateTable(){
        return new PlasticTable();
    }
    Chair* CreateChair(){
        return new PlasticChair();
    }
};


AbstractFurnitureFactory* AbstractFurnitureFactory::GetFactory(MaterialType type){
    switch (type)
    {
    case WOOD:
        return new WoodenFuniture();
    
    case PLASTIC:
        return new PlasticFuniture();
    
    default:
        cout<<"No supporting type = "<<type<<endl;
        return NULL;
    }
}

int main(){
    AbstractFurnitureFactory* factory = AbstractFurnitureFactory::GetFactory(PLASTIC);

    Chair* chair = factory->CreateChair();
    chair->CreateChair();

    Table* table = factory->CreateTable();
    table->CreateTable();

    return 0;
}

File: design_pattern/abtract_factory_design_pattern_table_chair.java
// https://gpcoder.com/4365-huong-dan-java-design-pattern-abstract-factory/

abstract class FurnitureAbstractFactory {
    abstract Chair createChair();
    abstract Table createTable();
}

class FurnitureFactory {
 
    private FurnitureFactory() {
 
    }
 
    // Returns a concrete factory object that is an instance of the
    // concrete factory class appropriate for the given architecture.
    public static FurnitureAbstractFactory getFactory(MaterialType materialType) {
        switch (materialType) {
        case FLASTIC:
            return new FlasticFactory();
        case WOOD:
            return new WoodFactory();
        default:
            throw new UnsupportedOperationException("This furniture is unsupported ");
        }
    }
}



class FlasticFactory extends FurnitureAbstractFactory {
 
    @Override
    public Chair createChair() {
        return new PlasticChair();
    }
 
    @Override
    public Table createTable() {
        return new PlasticTable();
    }
 
}

class WoodFactory extends FurnitureAbstractFactory {
 
    @Override
    public Chair createChair() {
        return new WoodChair();
    }
 
    @Override
    public Table createTable() {
        return new WoodTable();
    }
}

interface Chair {
    void create();
}
class PlasticChair implements Chair {
    @Override
    public void create() {
        System.out.println("Create plastic chair");
    }
}
class WoodChair implements Chair {
    @Override
    public void create() {
        System.out.println("Create wood chair");
    }
}
interface Table {
    void create();
}
class PlasticTable implements Table {
    @Override
    public void create() {
        System.out.println("Create plastic table");
    }
}
class WoodTable implements Table {
    @Override
    public void create() {
        System.out.println("Create wood table");
    }
}
enum MaterialType {
    FLASTIC, WOOD
}

class Client {
 
    public static void main(String[] args) {
 
        FurnitureAbstractFactory factory = FurnitureFactory.getFactory(MaterialType.FLASTIC);
 
        Chair chair = factory.createChair();
        chair.create(); // Create plastic chair
 
        Table table = factory.createTable();
        table.create(); // Create plastic table
    }
}